### PR TITLE
Feature/jpaley geom align

### PR DESCRIPTION
**Pull request by jonpaley**
This is a rather significant PR.

I have added a new "Align" package, which has a class that loads alignment constants and users can get a translation+rotation matrix from it.  There is an art::service for it too: "AlignService".  The idea is that these alignment constants will be in the global coordinate system, so they should be applied at that level.  They are currently being applied in the DetGeoMap class, which has a function that returns a line segment of a SSDCluster in the global coordinate system, but the Align class can be used anywhere.

I made a bunch of other changes to clean up bits of code here and there.  One thing probably worth pointing out is that I've further factorized services from their data providers.  In particular, the DetGeoMap service depends on the geometry and alignment [services], but instead of calling these service[s] within the DetGeoMap class, the DetGeoMap class is handed pointers to the Geometry and Align objects by the DetGeoMap service itself.  This removes the dependency of services within the data providers (a good thing in general), but it means that care must be taken to instantiate the services in the proper order and to reset pointers when necessary and appropriate.  This is all handled within the DetGeoMapService.

I have tested all of the code from simulation through Robert's single-track reconstruction module and the CAFMaker.  I've added an example of how to access track[segment] information in the CAF in an example script in CAFAna/example.  Everything works so far.  The alignment constants, currently set to 0 for all translations and rotations, live in the new ConstBase/Align directory.  We'll probably want to switch at some point to using a database.  And of course we need to test the usage of real alignment constants.

Reviewers should probably take some time to consider these changes.  I don't see any conflicts with the code in main.

### CODE REVIEW

**Review by lackey32 — COMMENTED** [review]
Thanks for all the code cleanup! It looks really nice.
Mostly minor comments.
And then my usual question: has this been tested on data?

File: Align/service/Align.fcl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+BEGIN_PROLOG
+
+standard_align:
+{
+  AbortIfFileNotFound: true
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply Align/service/Align.fcl:5]: Not sure about this particular parameter, but I envision others will be needed in the future, so this is basically a placeholder.

File: Align/Align.cxx [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    
+    ofile.close();
+    return true;
+  }
+  
+  //----------------------------------------------------------------------    
+
+  bool Align::LoadSSDConsts(std::string fname)
+  {
+    /*
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply Align/Align.cxx:148]: Ah, they should be removed.  I'll do that now-ish.

File: CAFMaker/CAFMaker_module.cc [hunk 3]
```diff
@@ -240,6 +243,21 @@ namespace caf {
     ssdhitsf.fLabel = fParams.SSDRawLabel();
     ssdhitsf.Fill(evt,rec);
 
+    // Get SpacePoints
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply CAFMaker/CAFMaker_module.cc:246]: These are raw digits, no reco job necessary.  Unless I've misunderstood the question?  [I note, however, that I really dislike having raw digits in the CAF!]

File: CAFMaker/CAFMaker.fcl [hunk 4]
```diff
@@ -7,13 +7,17 @@ standard_cafmaker:{
...
-    SSDClustLabel: "ssdclusts"
+    SSDClustLabel: "clust"
     BACkovHitLabel: "backovreco"
     GasCkovHitLabel: "gasckovreco"
-    //GetMCTruth:    true //use for creating MC CAF's
-    GetMCTruth:    false //use for creating data CAF's
-    GetMCHits:     false
+    GetMCTruth:    true //use for creating MC CAF's
+//    GetMCTruth:    false //use for creating data CAF's
+    GetMCHits:     true
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply CAFMaker/CAFMaker.fcl:15]: Probably...

File: DetGeoMap/service/DetGeoMapService.cc [hunk 5]
```diff
@@ -38,8 +44,13 @@ namespace emph
     //----------------------------------------------------------
     void DetGeoMapService::preBeginRun(const art::Run& run)
     {
+      std::cout << "DetGeoMapService::preBeginRun" << std::endl;
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply DetGeoMap/service/DetGeoMapService.cc:47]: Will remove

File: Geometry/service/Geometry_service.cc [hunk 6]
```diff
@@ -28,7 +28,7 @@ namespace emph
       art::ServiceHandle<runhist::RunHistoryService> rhs;
 
       reg.sPreBeginRun.watch(this, &GeometryService::preBeginRun);
-
+      //      reg.watchPreBeginRun(this, &GeometryService::preBeginRun);
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply Geometry/service/Geometry_service.cc:31]: done

File: CAFMaker/CAFMaker_module.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (lackey32) [reply CAFMaker/CAFMaker_module.cc:246]: Should have put this a little lower - I put it above what I was talking about instead. I'm talking about the Space Points and Track stuff.

**Comment by jonpaley** [comment]
> Thanks for all the code cleanup! It looks really nice. Mostly minor comments. And then my usual question: has this been tested on data?

I have not been able to test this on data.  I'm not sure how to do that until we are actually ready to start doing alignment work with data itself.  The changes only impact simulation and track reconstruction, which so far has only been applied to simulation.

File: CAFMaker/CAFMaker_module.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonpaley) [reply CAFMaker/CAFMaker_module.cc:246]: Ah, I see.  The job fhicl is in TrackReco/makesingletracks_job.fcl.  It works.

**Comment by lackey32** [comment]
> I have not been able to test this on data. I'm not sure how to do that until we are actually ready to start doing alignment work with data itself. The changes only impact simulation and track reconstruction, which so far has only been applied to simulation.
There were changes to some of the services. Though I would hope if those worked for MC they'd work for data too.

File: CAFMaker/CAFMaker_module.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (lackey32) [reply CAFMaker/CAFMaker_module.cc:246]: It needs adding to the CAFMaker job fcl though, right? So these can all be run in one job.

File: DetGeoMap/service/DetGeoMapService.cc [hunk 7]
```diff
@@ -22,6 +24,10 @@ namespace emph
 				       art::ActivityRegistry & reg):
       fUseGeometry (pset.get<bool>("UseGeometry"))
     {
+      //      art::ServiceHandle<runhist::RunHistoryService> rhs;
+      //      art::ServiceHandle<emph::geo::GeometryService> geo;
```
Comment (gavinsdavies) [inline DetGeoMap/service/DetGeoMapService.cc:28]: remove

File: DetGeoMap/service/DetGeoMapService.cc [hunk 8]
```diff
@@ -4,6 +4,8 @@
 //////////////////////////////////////////////////////////////////////////
 
 // EMPHATIC includes
+//#include "RunHistory/service/RunHistoryService.h"
```
Comment (gavinsdavies) [inline DetGeoMap/service/DetGeoMapService.cc:7]: remove

File: G4EMPH/BeamGen_module.cc [hunk 9]
```diff
@@ -275,15 +275,16 @@ namespace emph {
...
     TLorentzVector pos;
     
-    pos[2] = fZstart;
+    pos[2] = fZstart; // units are mm for this
+
     if (fXYHist) { // get random position from histogram
       fXYHist->GetRandom2(pos[0],pos[1]);
     }
     else { // get random position from flat or Gaussian distribution
       if (fXYDistSource == "FlatXY" || fXYDistSource == "flatXY" ||
```
Comment (gavinsdavies) [inline G4EMPH/BeamGen_module.cc:284]: make fhicl-configurable?

File: G4EMPH/BeamGen_module.cc [hunk 10]
```diff
@@ -329,7 +330,13 @@ namespace emph {
     
     // now add beam to the event    
     evt.put(std::move(beam));
-    
+    /*
+    std::cout << "Created beam particle with mom(" << pb[0]
```
Comment (gavinsdavies) [inline G4EMPH/BeamGen_module.cc:334]: could add to an mf::LogDebug

File: Geometry/Geometry.cxx [hunk 11]
```diff
@@ -110,23 +110,46 @@ namespace emph {
...
+      fIsLoaded(false), fGDMLFile(fname), fSSDStation(0)
     {
-      fGeoManager = 0;
+      fGeoManager = new TGeoManager("EMPHGeometry","EMPHATIC Geometry Manager");
+
       for ( int i = Trigger ; i < NDetectors ; i ++ ) fDetectorLoad[i] = false;
       fMagnetLoad = false;
-      this->SetGDMLFile(fname);
       fSSDSensorMap.clear();
+      std::cout << "Creating new Geometry object!" << std::endl << std::flush;
```
Comment (gavinsdavies) [inline Geometry/Geometry.cxx:151]: mf::LogInfo

File: StandardRecord/SRSpacePointBranch.h [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+////////////////////////////////////////////////////////////////////////
+#ifndef SRSPACEPOINTBRANCH_H
+#define SRSPACEPOINTBRANCH_H
+#include "StandardRecord/SRSpacePoint.h"
+
+#include <vector>
+
+namespace caf
+{
+  /// Reconstructed rings found by various algorithms
```
Comment (gavinsdavies) [inline StandardRecord/SRSpacePointBranch.h:12]: rings?

File: StandardRecord/SRSpacePointBranch.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    class SRSpacePointBranch
+    {
+    public:
+      SRSpacePointBranch();
+      ~SRSpacePointBranch();
+
+      std::vector<SRSpacePoint> sp;  ///< SpacePoints produced by ARichReco
+      size_t                nsp;     ///< number of arings
```
Comment (gavinsdavies) [inline StandardRecord/SRSpacePointBranch.h:20]: arings?

File: StandardRecord/SRTrackBranch.h [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+////////////////////////////////////////////////////////////////////////
+#ifndef SRTRACKBRANCH_H
+#define SRTRACKBRANCH_H
+#include "StandardRecord/SRTrack.h"
+
+#include <vector>
+
+namespace caf
+{
+  /// Reconstructed rings found by various algorithms
```
Comment (gavinsdavies) [inline StandardRecord/SRTrackBranch.h:12]: rings?

File: StandardRecord/SRTrackSegment.cxx [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+// \file    SRTrackSegment.cxx
+// \author  jpaley@fnal.gov
+// \brief   
+////////////////////////////////////////////////////////////////////////
+#include "StandardRecord/SRTrackSegment.h"
+
+namespace caf
+{
+  SRTrackSegment::SRTrackSegment() 
+  {
```
Comment (gavinsdavies) [inline StandardRecord/SRTrackSegment.cxx:11]: Add some NAN values so can just put default as 
`vtx.SetXYZ(0,0,0)`

```
vtx(std::numeric_limits<float>::signaling_NaN(),
        std::numeric_limits<float>::signaling_NaN(),
        std::numeric_limits<float>::signaling_NaN())
```

**Review by gavinsdavies — APPROVED** [review]
very nice, a lot of work and effort here. Thanks, @jonpaley !

File: Align/Align.cxx [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+#include <fstream>
+#include <cassert>
+#include <string>
+
+namespace emph {
+
+  Align::Align() 
+  {
+    fSSDMatrix.clear();
+    std::cout << "Created new Align object!" << std::endl;
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply Align/Align.cxx:19]: Better yet, I'm going to remove it, it's left over from debugging

File: Align/service/AlignService.cc [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  //----------------------------------------------------------
+    
+  AlignService::~AlignService()
+  {
+  }
+        
+  //----------------------------------------------------------
+  void AlignService::preBeginRun(const art::Run& )
+  {
+    std::cout << "AlignService::preBeginRun" << std::endl;
```
[the remark this answers is not in the document]
Comment (jonpaley) [reply Align/service/AlignService.cc:40]: Agreed, but removing altogether.

File: DetGeoMap/service/DetGeoMapService.cc (hunk 7 above)
Comment (jonpaley) [reply DetGeoMap/service/DetGeoMapService.cc:28]: will do

File: G4EMPH/BeamGen_module.cc (hunk 9 above)
Comment (jonpaley) [reply G4EMPH/BeamGen_module.cc:284]: "Zstart" is a fhicl-config.  Is that what you mean, or do you mean the units?

File: Geometry/Geometry.cxx (hunk 11 above)
Comment (jonpaley) [reply Geometry/Geometry.cxx:151]: Removed.

File: StandardRecord/SRSpacePointBranch.h (hunk 12 above)
Comment (jonpaley) [reply StandardRecord/SRSpacePointBranch.h:12]: Fixed

File: StandardRecord/SRSpacePointBranch.h (hunk 13 above)
Comment (jonpaley) [reply StandardRecord/SRSpacePointBranch.h:20]: Fixed

File: StandardRecord/SRTrackBranch.h (hunk 14 above)
Comment (jonpaley) [reply StandardRecord/SRTrackBranch.h:12]: Fixed

File: StandardRecord/SRTrackSegment.cxx (hunk 15 above)
Comment (jonpaley) [reply StandardRecord/SRTrackSegment.cxx:11]: Done.  Also set Mom elements to NaNs.

File: G4EMPH/BeamGen_module.cc (hunk 9 above)
Comment (gavinsdavies) [reply G4EMPH/BeamGen_module.cc:284]: ah ignore this one, i mis-read things.

**Pull request merged**